pkg/ingester/limiter_test.go

```diff
@@ -71,88 +71,125 @@ func runLimiterMaxFunctionTest(
-			expectedValue:            300,
+		"zone-awareness enabled, limit is enabled with replication-factor=3, some ingesters scaled down, and shard size 0": {
+			globalLimit:                 900,
+			ringReplicationFactor:       3,
+			ringIngesterCount:           9,
+			ringZoneAwarenessEnabled:    true,
+			ringZonesCount:              3,
+			healthyIngestersInZoneCount: 2, // in this zone ingesters are scaled down from 3 to 2
+			shardSize:                   0,
+			expectedValue:               450, // (900 / 3 zones / 2 healthy ingesters per zone) * 3 replication factor
+		},
+		"zone-awareness enabled, limit is enabled with replication-factor=3, all ingesters up and running, and shard size 5": {
+			globalLimit:                 900,
+			ringReplicationFactor:       3,
+			ringIngesterCount:           9,
+			ringZoneAwarenessEnabled:    true,
+			ringZonesCount:              3,
+			healthyIngestersInZoneCount: 3,   // 9 ingesters / 3 zones
+			shardSize:                   5,   // the expected number of ingesters per zone is (ceil(5/3) = 2
+			expectedValue:               450, // (900 / 3 zones / 2 healthy ingesters per zone) * 3 replication factor
+		},
+		"zone-awareness enabled, limit is enabled with replication-factor=3, some ingesters scaled down, and shard size 5": {
+			globalLimit:                 900,
+			ringReplicationFactor:       3,
+			ringIngesterCount:           9,
+			ringZoneAwarenessEnabled:    true,
+			ringZonesCount:              3,
+			healthyIngestersInZoneCount: 1,   // in this zone ingesters are scaled down from 3 to 1
+			shardSize:                   5,   // the expected number of ingesters per zone is (ceil(5/3) = 2
+			expectedValue:               900, // (900 / 3 zones / 1 healthy ingesters per zone) * 3 replication factor
+		},
+		"zone-awareness enabled, limit is enabled with replication-factor=3, all ingesters up and running, and the shard size > number of ingesters": {
+			globalLimit:                 900,
+			ringReplicationFactor:       3,
+			ringIngesterCount:           10,
+			ringZoneAwarenessEnabled:    true,
+			ringZonesCount:              3,
+			healthyIngestersInZoneCount: 3,   // 9 ingesters / 3 zones
+			shardSize:                   20,  // Greater than number of ingesters.
+			expectedValue:               300, // (900 / 3 zones / 3 healthy ingesters per zone) * 3 replication factor
+		},
+		"zone-awareness enabled, limit is enabled with replication-factor=3, some ingesters scaled down, and the shard size > number of ingesters": {
+			globalLimit:                 900,
+			ringReplicationFactor:       3,
+			ringIngesterCount:           10,
+			ringZoneAwarenessEnabled:    true,
+			ringZonesCount:              3,
+			healthyIngestersInZoneCount: 2,   // in this zone ingesters are scaled down from 3 to 2
+			shardSize:                   20,  // Greater than number of ingesters.
+			expectedValue:               450, // (900 / 3 zones / 2 healthy ingesters per zone) * 3 replication factor
 		},
 	}
```

Comment:
I'd suggest adding test cases for scaling-up too. That's also a scenario where limits are currently wrong.

Reply:
@pracucci has also mentioned the problem of scaling-up. I am not sure if this PR also fixes the scaling-up problem. Logically, I would say yes: we now know how many healthy ingesters are there in each zone, and we don't estimate that number like we used to do in the previous implementation. But I might miss some real-life use case. What do you think @pracucci @pstibrany?

Reply:
Yes, I think the problem is the same... when scaling up, say from 10+10+10 (30 ingesters in total) configuration to 15+10+10 (35 ingesters in total), previously this affected limits of ingesters in all zones. Now it will affect only scaled-up zone. I'd recommend adding a test showing that (one test for zone that's affected by scale up, one test for zone that's not affected by scale up).

Reply:
@pstibrany Shall I also add tests showing that limits do not change in the zones not affected by scaling down? Just for completeness

Reply:
Sounds good to me